File: android/app/src/main/java/com/example/campusai/data/remote/QrPayloadParser.kt
package com.example.campusai.data.remote

/**
 * QR payload 协议 — 跨端统一解析。
 *
 * 格式: campusmate://auth/web-login?v=1&sid=<session_id>&token=<scan_token>
 *
 * 不接受任意 URL，严格校验 scheme / host / path / version / 参数。
 * 纯 Kotlin 实现，不依赖 android.net.Uri，便于 JVM 单元测试。
 */
object QrPayloadParser {
    private const val SCHEME = "campusmate"
    private const val HOST = "auth"
    private const val PATH = "/web-login"
    private const val VERSION = 1

    data class QrPayload(
        val sessionId: String,
        val scanToken: String,
        val version: Int = VERSION,
    )

    /**
     * 解析并严格校验二维码字符串。
     * 返回 null 表示不是有效的 CampusMate 登录二维码。
     */
    fun parse(raw: String?): QrPayload? {
        if (raw.isNullOrBlank()) return null
        val trimmed = raw.trim()
        if (!trimmed.startsWith("$SCHEME://")) return null
        // 手动解析: campusmate://auth/web-login?v=1&sid=xxx&token=yyy
        val afterScheme = trimmed.substring("$SCHEME://".length)
        // 分离 authority+path 和 query
        val queryStart = afterScheme.indexOf('?')
        val authorityAndPath = if (queryStart >= 0) afterScheme.substring(0, queryStart) else afterScheme
        val queryString = if (queryStart >= 0) afterScheme.substring(queryStart + 1) else ""
        // 分离 host 和 path
        val pathStart = authorityAndPath.indexOf('/')
        val host = if (pathStart >= 0) authorityAndPath.substring(0, pathStart) else authorityAndPath
        val path = if (pathStart >= 0) authorityAndPath.substring(pathStart) else ""
        if (host != HOST) return null
        if (path != PATH) return null
        // 解析 query params
        val params = mutableMapOf<String, String>()
        for (pair in queryString.split("&")) {
            val eq = pair.indexOf('=')
            if (eq >= 0) {
                val key = pair.substring(0, eq)
                val value = pair.substring(eq + 1)
                params[key] = value
            }
        }
        val version = params["v"]?.toIntOrNull() ?: return null
        if (version != VERSION) return null
        val sessionId = params["sid"] ?: return null
        val scanToken = params["token"] ?: return null
        if (sessionId.length < 16) return null
        if (scanToken.length < 32) return null
        return QrPayload(sessionId = sessionId, scanToken = scanToken, version = version)
    }
}

File: wx/miniprogram/utils/qr-payload-parser.ts
/**
 * QR payload 协议 — 跨端统一生成与解析。
 *
 * 二维码内容格式:
 *   campusmate://auth/web-login?v=1&sid=<session_id>&token=<scan_token>
 *
 * 安全:
 * - 二维码只包含随机一次性凭据(session_id + scan_token)，不含 JWT / userId / 密码。
 * - browser_token 不写入二维码。
 * - 解析时严格校验 scheme / host / version / 参数格式，拒绝任意 URL。
 */

export const QR_PAYLOAD_SCHEME = 'campusmate'
export const QR_PAYLOAD_HOST = 'auth'
export const QR_PAYLOAD_PATH = '/web-login'
export const QR_PAYLOAD_VERSION = 1

export interface QrPayload {
  sessionId: string
  scanToken: string
  version: number
}

export function buildQrPayload(sessionId: string, scanToken: string): string {
  const params = `v=${QR_PAYLOAD_VERSION}&sid=${encodeURIComponent(sessionId)}&token=${encodeURIComponent(scanToken)}`
  return `${QR_PAYLOAD_SCHEME}://${QR_PAYLOAD_HOST}${QR_PAYLOAD_PATH}?${params}`
}

/**
 * 解析并严格校验二维码字符串。
 * 返回 null 表示不是有效的 CampusMate 登录二维码，不抛异常。
 */
export function parseQrPayload(raw: string | null | undefined): QrPayload | null {
  if (!raw || typeof raw !== 'string') return null
  const text = raw.trim()
  if (!text.startsWith(`${QR_PAYLOAD_SCHEME}://`)) return null

  const schemeEnd = text.indexOf('://')
  if (schemeEnd < 0) return null
  const afterScheme = text.slice(schemeEnd + 3)
  const pathStart = afterScheme.indexOf('/')
  if (pathStart < 0) return null

  const netloc = afterScheme.slice(0, pathStart)
  if (netloc !== QR_PAYLOAD_HOST) return null

  const afterHost = afterScheme.slice(pathStart)
  const queryStart = afterHost.indexOf('?')
  const pathPart = queryStart >= 0 ? afterHost.slice(0, queryStart) : afterHost
  if (pathPart !== QR_PAYLOAD_PATH) return null
  if (queryStart < 0) return null

  const queryStr = afterHost.slice(queryStart + 1)
  const params = parseQueryString(queryStr)
  const versionRaw = params['v']
  const sid = params['sid']
  const token = params['token']
  if (!versionRaw || !sid || !token) return null

  const version = Number(versionRaw)
  if (!Number.isInteger(version) || version !== QR_PAYLOAD_VERSION) return null
  if (sid.length < 16 || token.length < 32) return null

  return { sessionId: sid, scanToken: token, version }
}

function parseQueryString(q: string): Record<string, string> {
  const result: Record<string, string> = {}
  if (!q) return result
  for (const pair of q.split('&')) {
    if (!pair) continue
    const eq = pair.indexOf('=')
    if (eq < 0) {
      result[decodeURIComponent(pair)] = ''
      continue
    }
    const key = decodeURIComponent(pair.slice(0, eq))
    const value = decodeURIComponent(pair.slice(eq + 1))
    if (!(key in result)) result[key] = value
  }
  return result
}